src/Reopt/TypeInference/ConstraintGen.hs

```diff
@@ -129,165 +131,226 @@ newtype Warning = Warning String
 instance Show Warning where
   show (Warning w) = w
 
-type PhiType arch = Map (Some (FnPhiVar arch)) IType
+-- | Context available throughout all of the constraint generation
+newtype CGenGlobalContext arch = CGenGlobalContext
+  { _cgenMemory      :: Memory (ArchAddrWidth arch)
+  }
 
-data CGenContext arch = CGenContext {
-  -- Global state
-  cgenFunTypes        :: Map (ArchSegmentOff arch) (FunType arch)
-  -- | External functions (only named).
-  , cgenExtFunTypes   :: Map BSC.ByteString (FunType arch)
+makeLenses ''CGenGlobalContext
 
-  , cgenMemory        :: Memory (ArchAddrWidth arch)
-  -- Function state
-  , cgenCurrentFun    :: FunType arch
-  , cgenBlockPhiTypes :: Map (FnBlockLabel (ArchAddrWidth arch)) [IType]
+-- | Context available when generating constraints for a given module
+data CGenModuleContext arch = CGenModuleContext
+  { _cgenFunTypes    :: Map (ArchSegmentOff arch) (FunType arch)
+  , -- | External functions (only named).
+    _cgenExtFunTypes :: Map BSC.ByteString (FunType arch)
 
-  -- Current block state
-  , cgenCurrentPhis  :: PhiType arch
-}
+  -- (keep this last for convenient partial application)
+
+  , -- | Enclosing global context
+    _cgenGlobalContext :: CGenGlobalContext arch
+  }
+
+makeLenses ''CGenModuleContext
+
+-- | Context available when generating constraints for a given function
+data CGenFunctionContext arch = CGenFunctionContext
+  { -- | TODO
+    _cgenCurrentFun     :: FunType arch
+  , -- | Identifier for the current function
+    _cgenCurrentFunName :: BSC.ByteString
+  , -- |
+
+    -- TODO (val) could this be just TyVars?  Does it become obsolete when we
+    -- resolve TyVars?
+    _cgenBlockPhiTypes  :: Map (FnBlockLabel (ArchAddrWidth arch)) [IType]
+
+  -- (keep this last for convenient partial application)
+
+  , -- | Enclosing module context
+    _cgenModuleContext  :: CGenModuleContext arch
+  }
+
+makeLenses ''CGenFunctionContext
+
+-- | Context available when generating constraints for a given block.  At the
+-- moment, I managed to make it so that we don't need anything special, but it's
+-- nice to set this up for being future-proof.
+newtype CGenBlockContext arch = CGenBlockContext
+  {
+    -- (keep this last for convenient partial application)
+
+    -- | Enclosing function context
+    _cgenFunctionContext :: CGenFunctionContext arch
+  }
 
-data CGenState = CGenState {
+makeLenses ''CGenBlockContext
+
+data CGenState arch = CGenState {
     _nextFreeTyVar :: Int -- FIXME: use Nonce?
-  , _assignTypes   :: Map FnAssignId IType
+  , _assignTyVars  :: Map BSC.ByteString (Map FnAssignId TyVar)
+  , _tyVarsOrigin  :: Map TyVar (BSC.ByteString, FnAssignId)
+  , _tyVarTypes    :: Map TyVar IType
```

Comment:
Is this used?

Reply:
Together, `assignTyVars` and `tyVarTypes` allow us to go from a `(ByteString, FnAssignId)` pair to its inferred `IType` (through the associated `TyVar`). They decompose the old `Map FnAssignId IType` in two steps to have a convenient intermediate map from `FnAssignId` to `TyVar`.

`tyVarsOrigin` is used in the "Show constraints" feature to turn `TyVar`s back into their originating `FnAssignId`.